pkg/sql/sqlbase/index_encoding.go

```diff
@@ -239,35 +239,118 @@ func MakeSpanFromEncDatums(
 	return roachpb.Span{Key: startKey, EndKey: endKey}, containsNull, nil
 }
 
-// NeededColumnFamilyIDs returns a slice of FamilyIDs which contain
-// the families needed to load a set of neededCols
+// NeededColumnFamilyIDs returns the minimal set of column families required to
+// retrieve neededCols for the specified table and index. The returned FamilyIDs
+// are in sorted order.
 func NeededColumnFamilyIDs(
-	colIdxMap map[ColumnID]int, families []ColumnFamilyDescriptor, neededCols util.FastIntSet,
+	neededCols util.FastIntSet, table *TableDescriptor, index *IndexDescriptor,
 ) []FamilyID {
-	// Column family 0 is always included so we can distinguish null rows from
-	// absent rows.
-	needed := []FamilyID{0}
-	for i := range families {
-		family := &families[i]
+	// Copy neededCols so we can safely modify.
+	neededCols = neededCols.Copy()
+
+	// Build some necessary data structures for column metadata.
+	columns := table.ColumnsWithMutations(true)
```

Comment:
The `SpanBuilder` could cache these structures on creation of the object and pass them through to here.